src/ccrsa.c

```diff
@@ -41,91 +43,150 @@ const uint8_t *ccder_decode_rsa_priv(const ccrsa_full_ctx_t key, const uint8_t *
+ * Verify correctness of signature given. Digest is the digest of the data being signed.
  *
  * The return value is > 0 if there is an other error.
  * valid is set to false if there is an invalid signature.
  */
-int ccrsa_verify_pkcs1v15(ccrsa_pub_ctx_t key, const uint8_t *oid,
-                          size_t digest_len, const uint8_t *digest,
-                          size_t sig_len, const uint8_t *sig,
-                          bool *valid)
-{
-	*valid = 0;
+int ccrsa_verify_pkcs1v15(ccrsa_pub_ctx_t key, const uint8_t* oid, size_t digest_len, const uint8_t* digest, size_t sig_len, const uint8_t* sig, bool* valid) {
 #if DEBUG
 	printf("DARLING CRYPTO IMPL: %s\n", __PRETTY_FUNCTION__);
+
+	cc_println("key", ccrsa_pub_ctx_size(ccn_sizeof_n(ccrsa_ctx_n(key))), (const uint8_t*)ccrsa_ctx_zm(key).u);
+	cc_println("oid", strlen((const char*)oid), oid);
+	printf("digest_len: %zu\n", digest_len);
+	cc_println("digest", digest_len, digest);
+	printf("sig_len: %zu\n", sig_len);
+	cc_println("sig", sig_len, sig);
 #endif
 
-	printf("oid: %d\n", *oid);
+	*valid = false;
+
+	*valid = false;
 
-	printf("lengths: digest: %zu, signature: %zu\n", digest_len, sig_len);
+	cc_unit* s = NULL;
+	struct cczp* zp = NULL;
+	cc_unit* m = NULL;
+	uint8_t* em = NULL;
+	uint8_t* em_prime = NULL;
 
 	cc_size mod_size = ccrsa_ctx_n(key);
-	//cc_unit *modulus = *((cc_unit **)key.pub->b);
-	cc_unit *modulus = ccrsa_ctx_m(key);
-	cc_unit *exponent = ccrsa_ctx_e(key);
-	//cc_unit *exponent = *((cc_unit **)(key.pub->b+1));
+	cc_unit* modulus = ccrsa_ctx_m(key);
+	cc_unit* exponent = ccrsa_ctx_e(key);
 
+	// mod_size is how many units are allocated for the modulus
+	// mod_len is how many bytes (*bytes*, not units) are in use
+	cc_size mod_bits = ccn_bitlen(mod_size, modulus);
+	cc_size mod_len = (mod_bits + 7) / 8;
+	mod_size = ccn_nof(mod_bits);
 
-	// Maybe this should be determined dynamically
+	cc_size sig_bits = ccn_bitsof_size(sig_len);
 
-	// Length of signature should equal length of modulus
-	// mod size: 140198424041744
-	printf("mod size: %zu\n", mod_size);
-	if (sig_len != ccn_bitlen(mod_size, modulus))
-	{
 #if DEBUG
-		printf("not equal: siglen: %zu, modlen: %zu\n", sig_len, ccn_bitlen(mod_size, modulus));
+	printf("n of mod: %zu\n", mod_size);
+	printf("mod_len: %zu\n", mod_len);
+	cc_println("mod", ccn_sizeof_n(mod_size), (const uint8_t*)modulus);
+	cc_println("e", ccn_sizeof_n(mod_size), (const uint8_t*)exponent);
 #endif
-		return 1;
-	}
 
+	// number of bits in signature should equal the number of bits in modulus
+	if (sig_bits != mod_bits) {
+#if DEBUG
+		printf("%s: sig_bits (%zu) != mod_bits (%zu)\n", __PRETTY_FUNCTION__, sig_bits, mod_bits);
+#endif
+		goto fail;
+	}
 
 	// Compute big uint representation of signature
 	cc_size sig_units = ccn_nof_size(sig_len);
 	cc_size sig_bytes = ccn_sizeof_n(sig_units);
 
-	cc_unit *s = alloca(sig_bytes);
-	if (ccn_read_uint(sig_units, s, sig_len, sig))
-	{
+	s = malloc(sig_bytes);
```

Comment:
We are trying to avoid using malloc and instead use alloca so that it will work in the absence of a heap, like the original.

It's still used in some places but will be removed soon.

Reply:
Got it. I was just confused at first by the mix of `alloca` and `malloc` in the previous code so I defaulted to `malloc`.